system-helper/flatpak-system-helper.c

```diff
@@ -25,32 +25,186 @@
 
 static gboolean on_session_bus = FALSE;
 static gboolean no_idle_exit = FALSE;
 
 #define IDLE_TIMEOUT_SECS 10 * 60
+#define MAX_PULLS_PER_USER 16
+#define MOUNT_TIMEOUT 15
 
 /* This uses a weird Auto prefix to avoid conflicts with later added polkit types.
  */
 typedef PolkitAuthorizationResult AutoPolkitAuthorizationResult;
 typedef PolkitDetails             AutoPolkitDetails;
 typedef PolkitSubject             AutoPolkitSubject;
+typedef struct _FlatpakSystemHelperOngoingPull FlatpakSystemHelperOngoingPull;
 
 G_DEFINE_AUTOPTR_CLEANUP_FUNC (AutoPolkitAuthorizationResult, g_object_unref)
 G_DEFINE_AUTOPTR_CLEANUP_FUNC (AutoPolkitDetails, g_object_unref)
 G_DEFINE_AUTOPTR_CLEANUP_FUNC (AutoPolkitSubject, g_object_unref)
 
+struct _FlatpakSystemHelperOngoingPull
+{
+  FlatpakSystemHelper *object;
+  GDBusMethodInvocation *invocation;
+
+  const gchar *commit_id;
+  const gchar *ref;
+  guint watch_id;
+  uid_t uid;
+  guint64 current_timestamp;
+
+  gchar *root_folder_name;
+  gchar *user_folder_name;
+
+  GUnixMountMonitor *mount_monitor;
+  gulong mount_monitor_id;
+  guint unsuccessful_mount_timeout_id;
+  GSubprocess *bindfs;
+};
+
+static void
+system_helper_ongoing_pull_free (FlatpakSystemHelperOngoingPull *pull)
+{
+  g_autoptr(GFile) root_folder = g_file_new_for_path (pull->root_folder_name);
+  g_autoptr(GFile) user_folder = g_file_new_for_path (pull->user_folder_name);
+
+  flatpak_rm_rf (root_folder, NULL, NULL);
+  flatpak_rm_rf (user_folder, NULL, NULL);
+
+  g_clear_handle_id (&pull->watch_id, g_source_remove);
+
+  if (pull->bindfs != NULL)
+    {
+      if (umount (pull->user_folder_name) == -1)
+        {
+          int errsv = errno;
+          g_warning ("Failed to unmount bindfs dir: %s:%s, killing subprocess",
+                     g_strerror (errsv), pull->user_folder_name);
+          g_subprocess_force_exit (pull->bindfs);
+        }
+      g_clear_object (&pull->bindfs);
+   }
+
+  if (pull->mount_monitor_id)
+    {
+      g_signal_handler_disconnect (pull->mount_monitor, pull->mount_monitor_id);
+      pull->mount_monitor_id = 0;
+    }
+  g_object_unref (pull->mount_monitor);
+
+  g_clear_pointer (&pull->root_folder_name, g_free);
+  g_clear_pointer (&pull->user_folder_name, g_free);
+
+  g_ptr_array_remove (ongoing_pulls, pull);
+  g_ptr_array_unref (ongoing_pulls);
+  if (ongoing_pulls->len == 0)
+    ongoing_pulls = NULL;
+
+  g_slice_free (FlatpakSystemHelperOngoingPull, pull);
+}
+
+static void
+name_vanished_callback (GDBusConnection *connection, const gchar *name, gpointer user_data)
+{
+  FlatpakSystemHelperOngoingPull *pull = (FlatpakSystemHelperOngoingPull *) user_data;
+  system_helper_ongoing_pull_free (pull);
+}
+
+static void
+mounts_changed (GUnixMountMonitor *monitor, gpointer user_data)
+{
+  FlatpakSystemHelperOngoingPull *pull = (FlatpakSystemHelperOngoingPull *) user_data;
+  g_autoptr(GList) mounts = NULL;
+  GList *l = NULL;
+
+  mounts = g_unix_mounts_get (&pull->current_timestamp);
```

Comment:
The return value of this function is "transfer full" but this function only cleans up the GList container. It will leak all the `GUnixMountEntry *` values in the list.

Reply:
Addressed this in a172ca3f91aa5a003550a90de4dba0c6ab6b2441 with few others(which I have marked resolved)